wallet-new/src/Cardano/Wallet/Kernel/Compat.hs

```diff
@@ -0,0 +1,103 @@
+-- | This module exports various tools for compatibility with old wallet
+-- dependencies.
+
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE RankNTypes                 #-}
+
+module Cardano.Wallet.Kernel.Compat
+  ( DBReadT(DBReadT, unDBReadT)
+  , withMonadDBRead
+  , getCoreConfigurations
+  ) where
+
+import           Control.Monad.IO.Unlift (MonadUnliftIO, UnliftIO (UnliftIO),
+                     askUnliftIO, unliftIO, withUnliftIO)
+import           Control.Monad.Trans.Class (MonadTrans)
+import           Control.Monad.Trans.Reader (ReaderT (ReaderT), runReaderT)
+import           Control.Monad.Trans.Resource (transResourceT)
+import           Data.Conduit (transPipe)
+import           Pos.Core (BlockVersionData, CoreConfiguration,
+                     GeneratedSecrets, GenesisData, GenesisHash (GenesisHash),
+                     ProtocolConstants, ProtocolMagic, Timestamp,
+                     generatedSecrets, genesisBlockVersionData, genesisData,
+                     genesisHash, getGenesisHash, protocolConstants,
+                     withCoreConfiguration, withCoreConfigurations,
+                     withGeneratedSecrets, withGenesisBlockVersionData,
+                     withGenesisData, withGenesisHash, withProtocolConstants)
+import           Pos.Core.Configuration (HasConfiguration)
+import           Pos.DB.Class (MonadDBRead (..), Serialized (Serialized))
+import           Universum
+
+import           Pos.DB.Block (getSerializedBlock, getSerializedUndo)
+import           Pos.DB.Rocks.Functions (dbGetDefault, dbIterSourceDefault)
+import           Pos.DB.Rocks.Types (MonadRealDB, NodeDBs)
+
+--------------------------------------------------------------------------------
+
+-- | This monad transformer exists solely to provide a 'MonadRealDB' instance,
+-- as required by upstream libraries.
+newtype DBReadT m a = DBReadT { unDBReadT :: ReaderT NodeDBs m a }
+  deriving (Functor, Applicative, Monad, MonadThrow, MonadTrans, MonadIO)
+
+instance MonadUnliftIO m => MonadUnliftIO (DBReadT m) where
+  askUnliftIO =
+    DBReadT (withUnliftIO (\u -> pure (UnliftIO (unliftIO u . unDBReadT))))
+
+instance (HasConfiguration, MonadThrow (DBReadT m), MonadRealDB NodeDBs (ReaderT NodeDBs m))
```

Comment:
I'm very confused by the `MonadRealDB NodeDBs (ReaderT NodeDBs m))` constraint. You are requiring something for the _underlying_ (unwrapped) monad of `DBReadT`? I don't quite get it. Also, doesn't this mean that in order to use the `DBReadT` thing, I then am required, as a user, to somehow satisfy this `MonadRealDB` requirement -- for a monad I shouldn't even see, because it's an implementation detail of `DBReadT`?

Reply:
Yes, this is horrible. However, `DBReadT` is supposed to be a type internal to this module. I just removed it from the export list.